doc/BGP/BGP-supress-fib-pending.md

```diff
@@ -0,0 +1,804 @@
+Codes: K - kernel route, C - connected, S - static, R - RIP,
+       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
+       T - Table, v - VNC, V - VNC-Direct, A - Babel, F - PBR,
+       f - OpenFabric,
+       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
+       t - trapped, o - offload failure
+
+B>q 0.0.0.0/0 [20/0] via 10.0.0.1, PortChannel102, weight 1, 00:04:46
+  q                  via 10.0.0.5, PortChannel105, weight 1, 00:04:46
+  q                  via 10.0.0.9, PortChannel108, weight 1, 00:04:46
+  q                  via 10.0.0.13, PortChannel1011, weight 1, 00:04:46
+```
+
+Once zebra is notified about successfull route offload status the *offload* flag is set:
+
+```
+admin@sonic:~$ vtysh -c "show ip route 100.1.0.25/32 json"
+{
+    "100.1.0.25/32": [
+        {
+            "prefix": "100.1.0.25/32",
+            "prefixLen": 32,
+            "protocol": "bgp",
+            "vrfId": 0,
+            "vrfName": "default",
+            "selected": true,
+            "destSelected": true,
+            "distance": 20,
+            "metric": 0,
+            "installed": true,
+            "offloaded": true,
+            "table": 254,
+            "internalStatus": 80,
+            "internalFlags": 264,
+            "internalNextHopNum": 1,
+            "internalNextHopActiveNum": 1,
+            "nexthopGroupId": 4890,
+            "installedNexthopGroupId": 4890,
+            "uptime": "00:22:20",
+            "nexthops": [
+                {
+                    "flags": 3,
+                    "fib": true,
+                    "ip": "10.0.0.49",
+                    "afi": "ipv4",
+                    "interfaceIndex": 788,
+                    "interfaceName": "PortChannel1013",
+                    "active": true,
+                    "weight": 1
+                }
+            ]
+        }
+    ]
+}
+```
+
+#### 7.3. RouteOrch
+
+<!-- omit in toc -->
+#### ResponsePublisher in RouteOrch
+
+*RouteOrch* need to use existing *ResponsePublisher* API to be able to publish responses into *APPL_STATE_DB*.
+
+A snippet of *ResponsePublisher* API that is going to be used is given below.
+
+```c++
+void ResponsePublisher::publish(const std::string &table, const std::string &key,
+                                const std::vector<swss::FieldValueTuple> &intent_attrs,
+                                const ReturnCode &status,
+                                bool replace = false) override;
+```
+
+*RouteOrch* will not publish all fields and values to *APPL_STATE_DB* to reduce memory usage and reduce load on redis database. Instead, the key is published as well as the *protocol* field as this one is required for fpmsyncd to construct RTM_NEWROUTE message.
+
+Example usage in *RouteOrch*:
+
+```c++
+m_publisher.publish(APP_ROUTE_TABLE_NAME, kfvKey(kofvs), kfvFieldsValues(kofvs), ReturnCode(SAI_STATUS_SUCCESS), true);
+```
+
+Example data:
+```
+127.0.0.1:6379[14]> hgetall "ROUTE_TABLE:193.5.96.0/25"
```

Comment:
Why dont we have 'vrf' as the key in the ROUTE_TABLE? we could be supporting only 'default' VRF now, but it's important to have the VRF as the key for future enhancements with data VRF.

Reply:
This example is for default VRF. The key in APPL_STATE_DB is the same as in APPL_DB, with VRF included (if not default), so VRF is supported.

Reply:
ROUTE_TABLE in APPL_DB is legacy and we did not have VRF support originally and when we added the VRF support, we did not want to disrupt the existing schema and added the VRF key only for user VRF cases but APPL STATE_DB ROUTE_TABLE is a new schema, not sure why we still need to follow ROUTE_TABLE (APPL_DB) approach rather than having the VRF as the key by default even for 'default' VRF case as well. IMO. having VRF as one of the key by default is cleaner irrespective of default or user VRFs.

Reply:
Why ROUTE_TABLE in APPL_DB is legacy? Is there a new table with a new schema?

Here's the latest master and ROUTE_TABLE entry in APPL_DB:
```
127.0.0.1:6379> keys "ROUTE_TABLE:193.8.224.0/25"
1) "ROUTE_TABLE:193.8.224.0/25"
```

If it were a user created VRF route the key would be in a format <vrf>:<prefix>.
The idea of APPL_STATE_DB is to reflect the same schema that APPL_DB has.